src/res_core.ml

```diff
@@ -616,35 +616,30 @@ let parseHashIdent ~startPos p =
 let parseValuePath p =
   let startPos = p.Parser.startPos in
   let rec aux p path =
-    match p.Parser.token with
-    | Lident ident -> Longident.Ldot (path, ident)
-    | Uident uident ->
-      Parser.next p;
-      if p.Parser.token = Dot then (
-        Parser.expect Dot p;
-        aux p (Ldot (path, uident)))
-      else (
-        Parser.err p (Diagnostics.unexpected p.Parser.token p.breadcrumbs);
-        path)
-    | token ->
-      Parser.err p (Diagnostics.unexpected token p.breadcrumbs);
-      Longident.Ldot (path, "_")
+    let startPos = p.Parser.startPos in
+    let token = p.token in
+
+    Parser.next p;
+    if p.Parser.token = Dot then (
+      Parser.expect Dot p;
+
+      match p.Parser.token with
+      | Lident ident -> Longident.Ldot (path, ident)
+      | Uident uident -> aux p (Ldot (path, uident))
+      | token ->
+        Parser.err p (Diagnostics.unexpected token p.breadcrumbs);
+        Longident.Ldot (path, "_"))
+    else (
+      Parser.err p ~startPos ~endPos:p.prevEndPos (Diagnostics.lident token);
+      path)
   in
   let ident =
     match p.Parser.token with
     | Lident ident ->
       Parser.next p;
       Longident.Lident ident
     | Uident ident ->
-      Parser.next p;
-      let res =
-        if p.Parser.token = Dot then (
-          Parser.expect Dot p;
-          aux p (Lident ident))
-        else (
-          Parser.err p (Diagnostics.unexpected p.Parser.token p.breadcrumbs);
-          Longident.Lident ident)
-      in
+      let res = aux p (Lident ident) in
```

Comment:
I don't understand why a `Uident` has been turned into a `Lident`. The logic does not seem intuitive.
Is there a more intuitive way to express the logic?

Reply:
Oh I see it's for case +=.
That case is super rare hardly ever used.
I don't think it deserves a dedicated error message.

Reply:
As far as I know, it's because there is no `Uident` in `Longident`. I didn't touch it because it was originally using `Lident`.

Reply:
> Or are you targeting something more general?

`parseValuePath` is used not only in `parseTypeDefinitionOrExtension`, but also in many other places where parsing of the form `M1.M2.M3.t` is required. So I want to use it for general purpose by adding error handling for cases where the last item on the form is not 'Lident'.

If you think you need a more specific error message for type name, I'll work on it.

Reply:
Thank you for explaining.
It makes sense.

Reply:
> > I don't understand why a Uident has been turned into a Lident. The logic does not seem intuitive.
> 
> 
> 
> As far as I know, it's because there is no `Uident` in `Longident`. I didn't touch it because it was originally using `Lident`.

Before it was turning into Lident after an error, now before calling aux.
Which I'm not sure is even right.

Reply:
I think the logic has not changed since the `if-then-else` here has been moved to the `aux` function. If you think I didn't understand what you were talking about, could you please let me know again?